server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1675,13 +1677,16 @@ public void preRecovery(ActionListener<Void> listener) {
         indexEventListener.beforeIndexShardRecovery(this, indexSettings, listener);
     }
 
-    public void postRecovery(String reason) throws IndexShardStartedException, IndexShardRelocatedException, IndexShardClosedException {
-        synchronized (postRecoveryMutex) {
+    public void postRecovery(String reason, ActionListener<Void> listener) throws IndexShardStartedException, IndexShardRelocatedException,
+        IndexShardClosedException {
+        assert postRecoveryComplete == null;
```

Comment:
I was wondering if it would be better to use `SetOnce> for `postRecoveryComplete` if we want this check to happen in production, too.